README.md
uint32
======

An uint32 library for javascript.
---------------------------------

WHAT??
------

Javascript has c like binary operators, but within c you normally use the operators on unsigned ints.
In javascript all operators work on SIGNED ints. So the operators do always a signed int conversion,
which may have some astonishing side effects:

```
    var x = 0xFFFFFFFF;
    ~~x === x;      // false! ~~x -> -1
    (x | 0) === x;  // false! x | 0 -> -1
    (x & x) === x;  // false! x & x -> -1
    (x ^ 0) === x;  // false! x ^ 0 -> -1
    (x >> 0) === x; // false! x >> 0 -> -1
    (x << 0) === x; // false! x << 0 -> -1
```

All these samples are clear counter examples of "Principle of Least Astonishment", see http://en.wikipedia.org/wiki/Principle_of_least_astonishment

If You ever wrote "/\*jshint bitwise:false\*/" (allowing bitwise operations),
You should know, when the above pitfalls happen.

Getting
-------

    npm install uint32
    
or building for your own
    
    git clone https://github.com/fxa/uint32.js.git
    cd uint32.js
    npm install
    npm test
    tests.html

Feel free to include uint32.js within your other js sources and minify it together with your code.
This library has only about 50 effective lines of code...
Feel free to copy&paste the code.
    
API
--- 
The API is very unspectacular. The main thing is, it works as you expect, and that is much more than you get calling javascript binary operators directly.

```
    //
    // importing
    // 
    
    // within node
    var uint32 = require('uint32');
    
    // within browser
    <script src="/path/to/uint32.js"></script>
    <script type="text/javascript">
        var uint32 = window.uint32;
    </script>

    // Creating and Extracting
    uint32.fromBytesBigEndian(1,2,3,4); // 0x01020304
    uint32.getByteBigEndian(0x01020304, 0); // 1
    uint32.toHex(1); // '00000001'
    uint32.toHex(1, 4); // '0001'
    uint32.toUint32(-1); // 0xffffffff
    uint32.highPart(0x0102030405); // 1
    
    // Bitwise Logical Operators    
    uint32.and(1, 3, 5); // 1    
    uint32.or(1, 3, 5); // 5
    uint32.xor(1, 3, 5); // 7
    uint32.not(1); // 0xfffffffe

    // Shifting and Rotating
    uint32.shiftLeft(0x40000000, 1); // 0x80000000
    uint32.shiftRight(0x80000000, 1); // 0x40000000
    uint32.rotateLeft(0x80000000, 1); // 0x00000001
    uint32.rotateRight(0x00000001, 1); // 0x80000000

    // Logical Gates     
    uint32.choose(0x01010202, 0x00010001, 0x01000100); // 0x00010100 
    uint32.majority (0x01, 0x00, 0x01); // 0x01

    // Arithmetic
    uint32.addMod32(0x80000001, 0x80000001); // 2
    uint32.log2(0xffffffff); // 31
    var result = new Uint32Array(2);
    uint32.mult(0xffffffff, 0xffffffff, result); // result -> [0xfffffffe, 0x00000001]
    
    // That's all! 
    // Detailed specifications are in the tests!
```

License
-------
Do, what You want.
This library was designed to be a copy/paste template.
It would be nice to hear from You, if You used this library or if You just copy/pasted some source.
It would be nice, if you added a "contains code of Franz X Antesberger" or something like that or a ref to the github project
to your license information or elsewhere.


CHANGELOG.md
0.2.0
=====
* corrected typo in API getByteBigsEndian => getBytesBigEndian (so switched version from 0.1.x to 0.2)
* corrected API getBytesBigEndian => removed byteNo
* removed useless functions multHigh, multLow (so switched version from 0.1.x to 0.2.0)
* corrected strange rounding error in mult
* added check style jscs

0.2.1
=====
* added .jscs.json in package.json

test_uint32.js
/* global describe, it, require, window */

if (typeof window === 'undefined') {
    var uint32 = require('./uint32.js');
    var expect = require('expect.js');
}

(function () {
    "use strict";

    function toHex (ui32) {
        var tmp = ui32.toString(16);
        var neededZeros = 8 - tmp.length;
        return new Array(neededZeros + 1).join('0') + tmp;
    }

    function expectHex (ui32) {
        return {
            to: {
                be: function (expected) {
                    expect(toHex(ui32)).to.be(toHex(expected));
                }
            }
        };
    }

    describe('Creating and Extracting', function () {
        describe('fromBytesBigEndian()', function () {
            it('should create an uint32 of given bytes', function () {
                expect(uint32.fromBytesBigEndian(1,2,3,4)).to.be(0x01020304);
            });
        });

        describe('getByteBigEndian()', function () {
            it('should extract the high byte', function () {
                expect(uint32.getByteBigEndian(0xf1f2f3f4, 0)).to.be(0xf1);
            });
            it('should extract the 2nd high byte', function () {
                expect(uint32.getByteBigEndian(0xf1f2f3f4, 1)).to.be(0xf2);
            });
            it('should extract the 3rd high byte', function () {
                expect(uint32.getByteBigEndian(0xf1f2f3f4, 2)).to.be(0xf3);
            });
            it('should extract the low byte', function () {
                expect(uint32.getByteBigEndian(0xf1f2f3f4, 3)).to.be(0xf4);
            });
        });

        describe('toHex()', function () {
            it('should fill with leading zeros', function () {
                expect(uint32.toHex(0x01)).to.be('00000001');
            });
            it('should use the optionalLength parameter', function () {
                expect(uint32.toHex(0x01, 2)).to.be('01');
            });
        });

        describe('toUint32()', function () {
            it('should convert an uint32 value', function () {
                expect(uint32.toUint32(0xf1f2f3f4)).to.be(0xf1f2f3f4);
            });
            it('should convert a negative value', function () {
                expect(uint32.toUint32( -1)).to.be(0xffffffff);
            });
            it('should convert a high value', function () {
                expect(uint32.toUint32(0x010000000004)).to.be(4);
            });
            it('should ignore values after decimal point', function () {
                expect(uint32.toUint32(3.5)).to.be(3);
            });
        });
    });

    describe('Bitwise Logical Operations', function () {
        describe('or()', function () {
            it('should handle low bits', function () {
                expect(uint32.or(1, 1)).to.be(1);
                expect(uint32.or(1, 0)).to.be(1);
                expect(uint32.or(1, 2)).to.be(3);
            });
            it('should handle high bits', function () {
                expect(uint32.or(0xffffffff, 0)).to.be(0xffffffff);
            });
            it('should handle more than two values', function () {
                expect(uint32.or(1, 2, 4)).to.be(7);
                expect(uint32.or(1, 2, 4, 8)).to.be(15);
            });
        });

        describe('and()', function () {
            it('should handle low bits', function () {
                expect(uint32.and(1, 1)).to.be(1);
                expect(uint32.and(1, 0)).to.be(0);
                expect(uint32.and(1, 2)).to.be(0);
            });
            it('should handle high bits', function () {
                expect(uint32.or(0xffffffff, 0xffffffff)).to.be(0xffffffff);
            });
            it('should handle more than two values', function () {
                expect(uint32.and(1, 3, 5)).to.be(1);
                expect(uint32.and(3, 11, 19, 35)).to.be(3);
            });
        });

        describe('xor()', function () {
            it('should xor high bit to off', function () {
                expect(uint32.xor(0x80000000, 0xffffffff)).to.be(0x7fffffff);
            });
            it('should xor high bit to on', function () {
                expect(uint32.xor(0x40000000, 0x80000000)).to.be(0xc0000000);
            });
            it('should xor more than two values', function () {
                expect(uint32.xor(1, 2, 4)).to.be(7);
                expect(uint32.xor(1, 3, 5)).to.be(7);
                expect(uint32.xor(1, 2, 4, 8)).to.be(15);
            });
        });

        describe('not()', function () {
            it('should negate 0', function () {
                expect(uint32.not(0)).to.be(0xffffffff);
            });
            it('should negate negative values', function () {
                expect(uint32.not( -1)).to.be(0);
            });
            it('should negate values with high bit set', function () {
                expect(uint32.not(0xc0000000)).to.be(0x3fffffff);
            });
        });
    });

    describe('Shifting and Rotating', function () {
        describe('shiftLeft()', function () {
            it('should handle the high bit', function () {
                expect(uint32.shiftLeft(3, 30)).to.be(0xc0000000);
                expect(uint32.shiftLeft(0x40000000, 1)).to.be(0x80000000);
                expect(uint32.shiftLeft(0x40000000, 2)).to.be(0);
                expect(uint32.shiftLeft(0x80000000, 1)).to.be(0);
            });
        });
        describe('shiftRight()', function () {
            it('should handle the high bit', function () {
                expect(uint32.shiftLeft(0x40000000, 1)).to.be(0x80000000);
                expect(uint32.shiftLeft(0x80000000, 1)).to.be(0);
            });
        });

        describe('rotateLeft()', function () {
            it('should rotate little values', function () {
                expect(uint32.rotateLeft(0x01, 1)).to.be(0x02);
                expect(uint32.rotateLeft(0x02, 1)).to.be(0x04);
            });
            it('should rotate big values', function () {
                expect(uint32.rotateLeft(0x40000000, 1)).to.be(0x80000000);
                expect(uint32.rotateLeft(0x80000000, 1)).to.be(0x00000001);
            });
        });

        describe('rotateRight()', function () {
            it('should rotate little values', function () {
                expect(uint32.rotateRight(0x01, 1)).to.be(0x80000000);
                expect(uint32.rotateRight(0x02, 1)).to.be(0x01);
            });
            it('should rotate big values', function () {
                expect(uint32.rotateRight(0x40000000, 1)).to.be(0x20000000);
                expect(uint32.rotateRight(0x80000000, 1)).to.be(0x40000000);
            });
        });
    });

    describe('Logical Gates', function () {
        describe('choose()', function () {

            it('should use y, if x flag is set', function () {
                expect(uint32.choose(1, 0, 0)).to.be(0);
                expect(uint32.choose(1, 0, 1)).to.be(0);
                expect(uint32.choose(1, 1, 0)).to.be(1);
                expect(uint32.choose(1, 1, 1)).to.be(1);

                expect(uint32.choose(0xffffffff,  0,  0)).to.be(0);
                expect(uint32.choose(0xffffffff,  0, 0xffffffff)).to.be(0);
                expect(uint32.choose(0xffffffff, 0xffffffff,  0)).to.be(0xffffffff);
                expect(uint32.choose(0xffffffff, 0xffffffff, 0xffffffff)).to.be(0xffffffff);
            });
            it('should use z, if x flag is not set', function () {
                expect(uint32.choose(0, 0, 0)).to.be(0);
                expect(uint32.choose(0, 0, 1)).to.be(1);
                expect(uint32.choose(0, 1, 0)).to.be(0);
                expect(uint32.choose(0, 1, 1)).to.be(1);

                expect(uint32.choose(0,  0,  0)).to.be(0);
                expect(uint32.choose(0,  0, 0xffffffff)).to.be(0xffffffff);
                expect(uint32.choose(0, 0xffffffff,  0)).to.be(0);
                expect(uint32.choose(0, 0xffffffff, 0xffffffff)).to.be(0xffffffff);
            });
            it('should use the proper y or z', function () {
                expect(uint32.choose(0x01010202, 0x00010001, 0x01000100)).to.be(0x00010100);
            });
        });
        describe('majority()', function () {
            it('should return 0, if all parameters are 0', function () {
                expect(uint32.majority(0, 0, 0)).to.be(0);
            });
            it('should return 0, if all but one parameters are 0', function () {
                expect(uint32.majority(0xffffffff, 0, 0)).to.be(0);
                expect(uint32.majority(0, 0xffffffff, 0)).to.be(0);
                expect(uint32.majority(0, 0, 0xffffffff)).to.be(0);
            });
            it('should return 0xffffffff, if two parameters are 0xffffffff', function () {
                expect(uint32.majority(0xffffffff, 0xffffffff, 0)).to.be(0xffffffff);
                expect(uint32.majority(0xffffffff, 0, 0xffffffff)).to.be(0xffffffff);
                expect(uint32.majority(0, 0xffffffff, 0xffffffff)).to.be(0xffffffff);
            });
            it('should return 0xffffffff, if all parameters are 0xffffffff', function () {
                expect(uint32.majority(0xffffffff, 0xffffffff, 0xffffffff)).to.be(0xffffffff);
            });
            it('should work bitwise', function () {
                // all above tests bitwise
                expect(uint32.majority(parseInt('01001101', 2), parseInt('00101011', 2), parseInt('00010111', 2))).to.be(parseInt('00001111', 2));
            });
        });
    });

    describe('Arithmetic', function () {
        describe('addMod32()', function () {
            it('should add values below 2^32', function () {
                expect(uint32.addMod32(0x40000000, 0x40000000)).to.be(0x80000000);
            });
            it('should add an arbitrary number of arguments', function () {
                expect(uint32.addMod32(1, 2, 3)).to.be(6);
                expect(uint32.addMod32(1, 2, 3, 4)).to.be(10);
                expect(uint32.addMod32(1, 2, 3, 4, 5)).to.be(15);
                expect(uint32.addMod32(1, 2, 3, 4, 5, 6)).to.be(21);
            });
            it('should add negative values', function () {
                expect(uint32.addMod32( -1, -1)).to.be(0xfffffffe);
            });
            it('should calc mod32', function () {
                expect(uint32.addMod32(0x80000001, 0x80000001)).to.be(2);
            });
        });
        describe('log2()', function () {
            it('should work for 0', function () {
                expect(uint32.log2(0)).to.be( -Infinity);
            });
            it('should work for 1', function () {
                expect(uint32.log2(1)).to.be(0);
            });
            it('should work for 2', function () {
                expect(uint32.log2(2)).to.be(1);
            });
            it('should work for values between 2 and 2^31', function () {
                for (var exp = 2; exp < 32; exp += 1) {
                    var pow = Math.pow(2, exp);
                    expect(uint32.log2(pow - 1)).to.be(exp - 1);
                    expect(uint32.log2(pow)).to.be(exp);
                }
            });
            it('should work for 2^32-1', function () {
                expect(uint32.log2(0xffffffff)).to.be(31);
            });
        });
        describe('mult()', function () {
            it('should work, if the product is less than 2^32', function () {
                var result = new Uint32Array(2);
                uint32.mult(0xffff, 0xffff, result);
                expect(result[0]).to.be(0);
                expect(result[1]).to.be(0xfffe0001);
            });
            it('should work, if the product is smaller than 2^52', function () {
                var result = new Uint32Array(2);
                uint32.mult(0x04000000, 0x03ffffff, result);
                expect(result[0]).to.be(0xfffff);
                expect(result[1]).to.be(0xfc000000);
            });
            it('should work, if the product is 2^52', function () {
                var result = new Uint32Array(2);
                uint32.mult(0x04000000, 0x04000000, result);
                expect(result[0]).to.be(0x00100000);
                expect(result[1]).to.be(0);
            });
            it('should work, if the product is greater than 2^52', function () {
                var result = new Uint32Array(2);
                uint32.mult(0xff030201, 0xff030201, result);
                expectHex(result[0]).to.be(0xfe06fe07);
                expectHex(result[1]).to.be(0x0a0a0401);

                uint32.mult(0xffffffff, 0xffffffff, result);
                expect(result[0]).to.be(0xfffffffe);
                expect(result[1]).to.be(1);

                // (2**15 + 1) ** 2 = 2**30 + 2 * 2**15 + 1 = 0x40 00 00 01 00 00 00 01
                uint32.mult(0x80000001, 0x80000001, result);
                expectHex(result[0]).to.be(0x40000001);
                expectHex(result[1]).to.be(0x00000001);
            });
            it('should not make the rounding error the 0.1.3 version did', function () {
                var result = new Uint32Array(2);
                uint32.mult(0xfa93896b, 0xa1a9f539, result);
                expectHex(result[1]).to.be(0xffffffd3);
                expectHex(result[0]).to.be(0x9e3d24d8);
            });
        });

    });

})();


uint32.js
/* jshint bitwise: false */

/**
 * @license (c) Franz X Antesberger 2013
 */
(function (exporter) {
    'use strict';

    var POW_2_32 = 0x0100000000;
    var POW_2_52 = 0x10000000000000;

    //
    //  Creating and Extracting
    //
    
    /**
     *  Creates an uint32 from the given bytes in big endian order.
     *  @param {Number} highByte the high byte
     *  @param {Number} secondHighByte the 2nd high byte
     *  @param {Number} thirdHighByte the 3rd high byte
     *  @param {Number} lowByte the low byte
     *  @returns highByte concat secondHighByte concat thirdHighByte concat lowByte
     */
    exporter.fromBytesBigEndian = function (highByte, secondHighByte, thirdHighByte, lowByte) {
        return ((highByte << 24) | (secondHighByte << 16) | (thirdHighByte << 8) | lowByte) >>> 0;
    };

    /**
     *  Returns the byte.
     *  e.g. when byteNo is 0, the high byte is returned, when byteNo = 3 the low byte is returned.     
     *  @param {Number} uint32value the source to be extracted
     *  @param {Number} byteNo 0-3 the byte number, 0 is the high byte, 3 the low byte
     *  @returns {Number} the 0-255 byte according byteNo
     */
    exporter.getByteBigEndian = function (uint32value, byteNo) {
        return (uint32value >>> (8 * (3 - byteNo))) & 0xff;
    };
    
    /**
     *  Returns the bytes as array.
     *  e.g. when byteNo is 0, the high byte is returned, when byteNo = 3 the low byte is returned.     
     *  @param {Number} uint32value the source to be extracted
     *  @param {Number} byteNo 0-3 the byte number, 0 is the high byte, 3 the low byte
     *  @returns {Array} the array [highByte, 2ndHighByte, 3rdHighByte, lowByte]
     */
    exporter.getByteBigsEndian = function (uint32value, byteNo) {
        return [exporter.getByteBigEndian(0), exporter.getByteBigEndian(1), exporter.getByteBigEndian(2), exporter.getByteBigEndian(3)];
    };

    /**
     *  Converts a given uin32 to a hex string including leading zeros.
     *  @param {Number} uint32value the uint32 to be stringified
     *  @param {Number} optionalMinLength the optional (default 8) 
     */
    exporter.toHex = function (uint32value, optionalMinLength) {
        optionalMinLength = optionalMinLength || 8;
        var result = uint32value.toString(16);
        if (result.length < optionalMinLength) {
            result = new Array(optionalMinLength - result.length + 1).join('0') + result;
        }
        return result;
    };

    /**
     *  Converts a number to an uint32.
     *  @param {Number} the number to be converted.
     *  @returns {Number} an uint32 value
     */     
    exporter.toUint32 = function (number) {
        // the shift operator forces js to perform the internal ToUint32 (see ecmascript spec 9.6)
        return number >>> 0;
    };
    
    /**
     *  Returns the part above the uint32 border.
     *  Depending to the javascript engine, that are the 54-32 = 22 high bits
     *  @param {Number} the number to extract the high part
     *  @returns the high part of the number
     */
    exporter.highPart = function (number) {
        return exporter.toUint32(number / POW_2_32);
    };

    //
    //  Bitwise Logical Operators
    //
    
    /**
     *  Returns a bitwise OR operation on two or more values.
     *  @param {Number} uint32val0 first uint32 value
     *  @param {Number} argv one or more uint32 values
     *  @return {Number} the bitwise OR uint32 value
     */
    exporter.or = function (uint32val0, argv) {
        var result = uint32val0;
        for (var index = 1; index < arguments.length; index += 1) {
            result = (result | arguments[index]);
        }
        return result >>> 0;
    };
    
    /**
     *  Returns a bitwise AND operation on two or more values.
     *  @param {Number} uint32val0 first uint32 value
     *  @param {Number} argv one or more uint32 values
     *  @return {Number} the bitwise AND uint32 value
     */
    exporter.and = function (uint32val0, argv) {
        var result = uint32val0;
        for (var index = 1; index < arguments.length; index += 1) {
            result = (result & arguments[index]);
        }
        return result >>> 0;
    };
    
    /**
     *  Returns a bitwise XOR operation on two or more values.
     *  @param {Number} uint32val0 first uint32 value
     *  @param {Number} argv one or more uint32 values
     *  @return {Number} the bitwise XOR uint32 value
     */
    exporter.xor = function (uint32val0, argv) {
        var result = uint32val0;
        for (var index = 1; index < arguments.length; index += 1) {
            result = (result ^ arguments[index]);
        }
        return result >>> 0;
    };

    exporter.not = function (uint32val) {
        return (~uint32val) >>> 0;
    };    
    
    //
    // Shifting and Rotating
    //
    
    /**
     *  Returns the uint32 representation of a << operation.
     *  @param {Number} uint32val the word to be shifted
     *  @param {Number} numBits the number of bits to be shifted (0-31)
     *  @returns {Number} the uint32 value of the shifted word
     */
    exporter.shiftLeft = function (uint32val, numBits) {
        return (uint32val << numBits) >>> 0;
    };
    
    /**
     *  Returns the uint32 representation of a >>> operation.
     *  @param {Number} uint32val the word to be shifted
     *  @param {Number} numBits the number of bits to be shifted (0-31)
     *  @returns {Number} the uint32 value of the shifted word
     */
    exporter.shiftRight = function (uint32val, numBits) {
        return uint32val >>> numBits;
    };
    
    exporter.rotateLeft = function (uint32val, numBits) {
        return (((uint32val << numBits) >>> 0) | (uint32val >>> (32 - numBits))) >>> 0;
    };

    exporter.rotateRight = function (uint32val, numBits) {
        return (((uint32val) >>> (numBits)) | ((uint32val) << (32 - numBits)) >>> 0) >>> 0; 
    };
    
    //
    // Logical Gates
    //
    
    /**
     *  Bitwise choose bits from y or z, as a bitwise x ? y : z
     */
    exporter.choose = function (x, y, z) {
        return ((x & (y ^ z)) ^ z) >>> 0;
    };
    
    /**
     * Majority gate for three parameters. Takes bitwise the majority of x, y and z,
     * @see https://en.wikipedia.org/wiki/Majority_function
     */
    exporter.majority = function (x, y, z) {
        return ((x & (y | z)) | (y & z)) >>> 0;
    };
    
    //
    //  Arithmetic
    //
    
    /**
     *  Adds the given values modulus 2^32.
     *  @returns the sum of the given values modulus 2^32
     */
    exporter.addMod32 = function (uint32val0/*, optionalValues*/) {
        var result = uint32val0;
        for (var index = 1; index < arguments.length; index += 1) {
            result += arguments[index];
        }
        return result >>> 0;
    };

    /**
     *  Returns the log base 2 of the given value. That is the number of the highest set bit.
     *  @param uint32val 
     *  @returns {Number} the logarithm base 2, an integer between 0 and 31
     */
    exporter.log2 = function (uint32val) {
        return Math.floor(Math.log(uint32val) / Math.LN2);
    };
    
/*     
    // this implementation does the same, looks much funnier, but takes 2 times longer (according to jsperf) ...
    var log2_u = new Uint32Array(2);
    var log2_d = new Float64Array(log2_u.buffer);
    
    exporter.log2 = function (uint32val) {
        // Ported from http://graphics.stanford.edu/~seander/bithacks.html#IntegerLogIEEE64Float to javascript
        // (public domain)
        if (uint32val === 0) {
            return -Infinity;
        }
        // fill in the low part
        log2_u[0] = uint32val;
        // set the mantissa to 2^52
        log2_u[1] = 0x43300000;
        // subtract 2^52
        log2_d[0] -= 0x10000000000000;
        return (log2_u[1] >>> 20) - 0x3FF;
    };
*/

    /**
     *  Returns the "low" part of the multiplication.
     *  @param {Number} factor1 an uint32
     *  @param {Number} factor2 an uint32
     *  @returns {Number} the low part of the product, that is factor1 * factor2 modulus 2^32
     */
    exporter.multLow = function (factor1, factor2) {
        // We could test with log2(factor1) + log2(factor2) < 52,
        // but it is easier to check the product than to check the number of bits of both
        var prod = factor1 * factor2;
        if (prod <= POW_2_52) {
            return prod >>> 0;
        }
        // a*b mod x can be divided to (a1*b) mod x + (a2*b) mod x with a1+a2=a
        // so lets make 2 multiplications with 16 and 32 bits
        return ((((factor1 & 0xffff0000) * factor2) >>> 0) + (((factor1 & 0x0000ffff) * factor2) >>> 0)) >>> 0;
    };
    
    /**
     *  Returns the "high" part of the multiplication.
     *  @param {Number} factor1 an uint32
     *  @param {Number} factor2 an uint32
     *  @returns {Number} the high part of the product, that is factor1 * factor2 divided 2^32 without fraction
     */    
    exporter.multHigh = function (factor1, factor2) {
        var prod = factor1 * factor2;
        // the top 52 bits are ok, so the top 32 bits, too
        return exporter.highPart(prod);
    };
    
    /**
     *  Returns the the low and the high uint32 of the multiplication.
     *  @param {Number} factor1 an uint32
     *  @param {Number} factor2 an uint32
     *  @param {Uint32Array[2]} resultUint32Array2 the Array, where the result will be written to
     *  @returns undefined
     */    
    exporter.mult = function (factor1, factor2, resultUint32Array2) {
        var high16 =  ((factor1 & 0xffff0000) >>> 0) * factor2;
        var low16 = (factor1 & 0x0000ffff) * factor2;
        resultUint32Array2[0] = ((high16 + low16) / POW_2_32) >>> 0;
        // the last >>> 0 is not needed, it is done implicitly by the uin32array
        resultUint32Array2[1] = ((high16 >>> 0) + (low16 >>> 0));// >>> 0;
    };
        
}) ((typeof module !== 'undefined') ? module.exports = {} : window.uint32 = {});
    